File: C#/0234. Palindrome Linked List.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public bool IsPalindrome(ListNode head) {
        ListNode node=new ListNode(0);
        node.next=head;
        ListNode pslow=node;
        ListNode pfast=node;
        while (pfast.next!=null && pfast.next.next!=null){
            pslow=pslow.next;
            pfast=pfast.next.next;
        }
        ListNode newhead;
        if(pfast.next!=null){
            newhead=pslow.next.next;
        }
        else{
            newhead=pslow.next;
        }
        
        //reverse linked list
        Solution rep=new Solution();
        newhead=rep.reverseList(newhead);
        
        //compare
        while(newhead!=null){
            if(newhead.val!=head.val){
                return false;
            }
            else{
                newhead=newhead.next;
                head=head.next;
            }
        }
        return true;
        
    }
    
    public ListNode reverseList(ListNode head){
        ListNode prev=null;
        while (head!=null){
            ListNode cur=head;
            head=head.next;
            cur.next=prev;
            prev=cur;
        }
        return prev;
    }
}

File: archives/20190519python/0234. Palindrome Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def isPalindrome(self, head: ListNode) -> bool:
        if not head or not head.next:
            return True
        pslow=head
        pfast=head
        while pfast.next and pfast.next.next:
            pslow=pslow.next
            pfast=pfast.next.next
        return self.isSubPalindrome(head,pslow.next)
    
    def isSubPalindrome(self, headA, headB):
        headB=self.reverseList(headB)
        while headB:
            if headA.val==headB.val:
                headA=headA.next
                headB=headB.next
            else:
                return False
        return True
        
    def reverseList(self,head):
        prev=None
        while head:
            cur=head
            head=head.next
            cur.next=prev
            prev=cur
        return prev
